### DISCUSSION TITLE
What is faster than for loop in resultproxy object?

### DISCUSSION THREAD

**Discussion by poppyc0ck** [sqlalchemy/sqlalchemy#8543] [Usage Questions] [answered]
I got the 20,000 records in resultproxy object.
`resultproxy = conn.execute('SELECT * FROM TABLE')`

but, it is so slow for forloop that covert to dict (or serialization)
`dict = [dict(x) for x in resultproxy]`

method fetchall is the same result.

**Comment by CaselIT** [comment]
hi,

without a bit more context it's a bit hard to give suggestions. Like database, sqlalcehmy version etc
What is slow? the execution of the query or the conversion to dict?
Also can you quantify "slow"?

**Reply by poppyc0ck** [reply]
Good day, @CaselIT  Sry about that.
database : IBM DB2 11(via ibm_db_sa dirver)
sqlalcehmy version: 1.3.23(tried 1.4.41)

It's is fast that got the resultproxy object of query. 
It's slow that loops each iteration of row(rowproxy),150 seconds for over 20,000 records, 

In other words, I don't how to iterate rowproxy quicky or convert resultproxy to dict without iteration.

**Reply by zzzeek** [reply]
Have you measured that the "conversion to dict" is the issue, and not just the speed of your database and/or some issues within the DB2 database driver?

**Comment by zzzeek** [comment, marked as answer]
Here's how we measure SQLAlchemy's part of this, we use a SQLite memory database which has extremely low overhead.    Any time spent outside of what you would see in this test is a function of either A. database/DB2 driver / network slowness  or B. you have extremely large amounts of data in each row which takes a long time over the network.

As for the Python side, this test illustrates that for 20K rows, it makes basically no difference.  Try using this suite with your database:

```py
import random
import string
import time

from sqlalchemy import create_engine


pool_of_random_words = [
    "".join(random.choice(string.ascii_lowercase) for i in range(200))
    for word in range(1000)
]


e = create_engine("sqlite://")

with e.connect() as conn:
    conn.exec_driver_sql(
        "create table my_table "
        "(id integer primary key, x varchar, y varchar, z varchar)"
    )
    conn.exec_driver_sql(
        "INSERT INTO my_table (x, y, z) VALUES (?, ?, ?)",
        [
            (
                random.choice(pool_of_random_words),
                random.choice(pool_of_random_words),
                random.choice(pool_of_random_words),
            )
            for i in range(20000)
        ],
    )
    print("INSERTED 20000 RECS OF 600 CHARACTERS EACH!\n\n")

    def fmany(r):
        result = []
        while True:
            batch = r.fetchmany(200)
            if not batch:
                break
            result.extend(batch)

        return result

    def fone(r):
        result = []
        while True:
            row = r.fetchone()
            if row is not None:
                result.append(row)
            else:
                break

        return result

    for name, meth in [
        (".all()", lambda r: r.all()),
        (".__iter__()", lambda r: [row for row in r]),
        (".fetchmany(200)", fmany,),
        (".fetchone()", fone,),
    ]:
        result = conn.exec_driver_sql("SELECT * FROM my_table")
        print(f"ABOUT TO FETCH 20K ROWS USING {name}")
        now = time.time()
        rows = meth(result)
        total = time.time() - now
        print(f"AND THAT TOOK...{total} SECONDS!\n\n")
        assert len(rows) == 20000

```

output for SQLite, we can see all options take about 60 milliseconds to fetch 20K rows with no database overhead:

```
INSERTED 20000 RECS OF 600 CHARACTERS EACH!


ABOUT TO FETCH 20K ROWS USING .all()
AND THAT TOOK...0.06931281089782715 SECONDS!


ABOUT TO FETCH 20K ROWS USING .__iter__()
AND THAT TOOK...0.056931495666503906 SECONDS!


ABOUT TO FETCH 20K ROWS USING .fetchmany(200)
AND THAT TOOK...0.052767276763916016 SECONDS!


ABOUT TO FETCH 20K ROWS USING .fetchone()
AND THAT TOOK...0.06192374229431152 SECONDS!
```

**Reply by poppyc0ck** [reply]
@zzzeek
u'r right. The root cause is network slowness. I defined the problem in the wrong direction.
Thanks for getting back to me. have a nice day :)
